File: Mizuki/src/frames.zig
// SPDX-License-Identifier: MPL-2.0

const c = @import("c.zig").c;

/// Frame containing the base video
pub const VideoFrame = extern struct {
    frame_number: c_int = -1,
    timestamp: c_longlong = 0,
    width: c_int,
    height: c_int,
    pitch: c_int,
    flipped: c_int,
    data: [*]u8, // ffms_Frame.Data[0]
    valid: c_int = 0, // bool
};

/// Frame containing rendered subtitles on a transparent background
pub const SubtitleFrame = extern struct {
    hash: u32 = 0,
    timestamp: c_longlong = 0,
    width: c_int,
    height: c_int,
    pitch: c_int,
    flipped: c_int,
    data: [*]u8,
    valid: c_int = 0, // bool
};

/// Frame containing audio samples
pub const AudioFrame = extern struct {
    data: [*]i16,
    length: i64,
    channel_count: c_int,
    sample_count: i64,
    sample_rate: c_int,
    duration_ms: i64,
    valid: c_int = 0, // bool
};

/// Collection of the VideoFrame and SubtitleFrame
pub const FrameGroup = extern struct {
    video_frame: *VideoFrame,
    subtitle_frame: *SubtitleFrame,
};

/// Generic bitmap
pub const Bitmap = extern struct {
    width: c_int,
    height: c_int,
    pitch: c_int,
    data: [*]u8,
};

File: Holo/Media/Providers/Frames.cs
// SPDX-License-Identifier: MPL-2.0

using System.Runtime.InteropServices;

namespace Holo.Media.Providers;

[StructLayout(LayoutKind.Sequential)]
public unsafe struct FrameGroup
{
    public VideoFrame* VideoFrame;
    public SubtitleFrame* SubtitleFrame;
}

[StructLayout(LayoutKind.Sequential)]
public struct VideoFrame
{
    public int FrameNumber;
    public long Timestamp;
    public int Width;
    public int Height;
    public int Pitch;
    public int Flipped;
    public unsafe byte* Data;
    public int Valid;
}

[StructLayout(LayoutKind.Sequential)]
public struct SubtitleFrame
{
    public int FrameNumber;
    public long Timestamp;
    public int Width;
    public int Height;
    public int Pitch;
    public int Flipped;
    public unsafe byte* Data;
    public int Valid;
}

[StructLayout(LayoutKind.Sequential)]
public struct AudioFrame
{
    public unsafe short* Data;
    public long Length;
    public int ChannelCount;
    public long SampleCount;
    public int SampleRate;
    public long DurationMillis;
    public int Valid;
}

[StructLayout(LayoutKind.Sequential)]
public struct TrackInfo
{
    public int Index;
    public string Codec;
    // public string Language;
    // public string Title;
}

[StructLayout(LayoutKind.Sequential)]
public struct Bitmap
{
    public int Width;
    public int Height;
    public int Pitch;
    public unsafe byte* Data;
}
